File: 백준/Silver/24479. 알고리즘 수업 － 깊이 우선 탐색 1/알고리즘 수업 － 깊이 우선 탐색 1.py
import sys
sys.setrecursionlimit(150000)
input = sys.stdin.readline

v = 1
def dfs(t):
    global v
    visited[t] = v
    graph[t].sort()
    for g in graph[t]:
        if visited[g]==0:
            v += 1
            dfs(g)

# n은 정점 집합 m은 간선 집합 r은 시작 정점
n, m, r = map(int, input().split())
graph = [[]*n for _ in range(n+1)]
visited = [0]*(n+1)
for _ in range(m):
    a, b = map(int, input().split())
    graph[a].append(b)
    graph[b].append(a)

# 방문
dfs(r)
for i in range(1, n + 1):
    print(visited[i])

File: 백준/Silver/24479. 알고리즘 수업 － 깊이 우선 탐색 1/알고리즘 수업 － 깊이 우선 탐색 1.java
import java.util.*;
import java.io.*;

public class Main {
    static int n, m;
    static int[] visited;
    static ArrayList<ArrayList<Integer>> arr;
    static int count = 0;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());
        int r = Integer.parseInt(st.nextToken());
        visited = new int[n+1];
        arr = new ArrayList<>();
        for (int i = 0; i<=n; i++) {
            arr.add(new ArrayList<>());
        }
        for (int i = 0; i<m; i++) {
            st = new StringTokenizer(br.readLine());
            int s = Integer.parseInt(st.nextToken());
            int e = Integer.parseInt(st.nextToken());
            arr.get(s).add(e);
            arr.get(e).add(s);
        }

        for (int i = 0; i<arr.size(); i++) {
            Collections.sort(arr.get(i));
        }

        count = 1;
        visited[r] = count;
        dfs(r);

        for (int i = 1; i<=n; i++) {
            System.out.println(visited[i]);
        }
    }

    public static void dfs(int now) {
        for (int i: arr.get(now)) {
            if (visited[i]==0) {
                count++;
                visited[i] = count;
                dfs(i);
            }
        }
    }
}